### PR TITLE
RND-33794 |  HealthKit background delivery never fired

### PR DESCRIPTION
**Pull request by giladna**
## Description

- Adds support for passing a raw number of seconds to `registerHealthBackgroundSync({ syncInterval })`, removing the 1-hour minimum enforced by the previous string-only API
- Updates the Hello-Heart/react-native-health fork with matching native + type changes

### @helloheart/core (this repo)

**src/health/backgroundSync.ts**
- `syncInterval` union extended with `number` (seconds)
- `'every1minute'` alias removed — pass `60` directly instead
- JSDoc updated: documents float rounding, valid range, and 86400s fallback for invalid input

**src/health/__tests__/backgroundSync.test.ts**
- 5 new tests: numeric pass-through (60), float pass-through (1.5), edge cases (0, -1), string alias unchanged
- 10 tests total, all passing

**docs/health/background-sync-flexible-interval.md** — new feature doc

**package.json / yarn.lock** — pinned to `react-native-health#RND-33794`

---

### Hello-Heart/react-native-health (RND-33794 branch)

**RCTAppleHealthKit/RCTAppleHealthKit.m**
- `configureBackgroundSync` reads `syncInterval` as `id` — branches on `NSNumber` (raw seconds) vs `NSString` (named alias)
- Numeric values: floats rounded to nearest integer via `MAX(1.0, round(raw))`; zero, negative, NaN, Infinity fall back to `86400s` (24h default)
- All three invalid-input paths (number, string, wrong type) consistently fall back to `86400s`
- Replaced `dispatch_once` lazy-init guard with `_observersInitialized` BOOL ivar — nil-bridge first calls no longer permanently block observer registration
- Removed `every1minute` dead-code branch from `syncIntervalFromString`
- Scoped `NSNotificationCenter` removal in `startObserving` to HealthKit event names only — prevents silently dropping unrelated observers

**RCTAppleHealthKit/RCTAppleHealthKit+Methods_Fitness.m**
- Guard against nil `sampleType` in `fitness_registerObserver`
- Fix: `MindfulSession` registered as `HKCategoryType` observer (was silently skipped — background delivery never fired)

**index.d.ts** — `SyncInterval` extended with `| number`; JSDoc documents rounding and fallback behavior

**docs/configureBackgroundSync.md** — updated with numeric usage, float rounding, and input validation details

## Usage

```js
// Testing — any interval in seconds
registerHealthBackgroundSync({ enabled: true, syncInterval: 60 })   // 1 min
registerHealthBackgroundSync({ enabled: true, syncInterval: 300 })  // 5 min

// Production — named aliases unchanged
registerHealthBackgroundSync({ enabled: true, syncInterval: 'every6hours' })
```

## Test Plan

- Run `yarn test src/health/__tests__/backgroundSync.test.ts` — 10/10 pass
- iOS rebuild required after native change (`pod install` + clean build)
- Verify `[HealthKit] Background sync enabled, interval 60s` appears in Xcode console after calling with `syncInterval: 60`
- Verify observer fires within ~1 minute by adding a health sample on device

## Before merging to main

- Tag `react-native-health` branch as a release (e.g. `v1.22.0`)
- Update `package.json` from `#RND-33794` to `#v1.22.0`
- Re-run `yarn install` to update `yarn.lock`

### CODE REVIEW

**Review by giladna — COMMENTED** [review]
Overall looks solid. Two things worth addressing:

File: RCTAppleHealthKit/RCTAppleHealthKit.m [hunk 1]
```diff
@@ -1101,12 +1130,13 @@ - (void)initializeBackgroundObservers:(RCTBridge *)bridge
 // Will be called when this module's first listener is added.
 -(void)startObserving {
     NSNotificationCenter *center = [NSNotificationCenter defaultCenter];
-        for (NSString *notificationName in [self supportedEvents]) {
-            [center addObserver:self
-                   selector:@selector(emitEventInternal:)
-                       name:notificationName
-                     object:nil];
-        }
+    [center removeObserver:self];
```
Comment (giladna) [inline RCTAppleHealthKit/RCTAppleHealthKit.m]: \`[center removeObserver:self]\` removes **all** NSNotificationCenter observers for this instance, not just the health-kit event names. If the class ever registers observers outside of \`startObserving\` (e.g. app-lifecycle or bridge notifications), they'll be silently dropped each time the first JS listener is added.

Safer to remove only the names you're about to re-add:

\`\`\`objc
for (NSString *notificationName in [self supportedEvents]) {
    [center removeObserver:self name:notificationName object:nil];
    [center addObserver:self
           selector:@selector(emitEventInternal:)
               name:notificationName
             object:nil];
}
\`\`\`

File: index.d.ts [hunk 2]
```diff
@@ -936,14 +936,18 @@ declare module 'react-native-health' {
...
+   * Pass a named alias or a raw number of **seconds** (e.g. `60` for 1 minute).
+   */
   export type SyncInterval =
     | 'every1hour'
     | 'every6hours'
     | 'every12hours'
     | 'every24hours'
     | 'every48hours'
     | 'everyweek'
+    | number
```
Comment (giladna) [inline index.d.ts]: \`| number\` accepts \`0\`, negative values, \`NaN\`, and \`Infinity\` — all of which get clamped to \`1s\` natively. TypeScript can't enforce positive-finite-number natively, but a JSDoc note on the \`BackgroundSyncOptions.syncInterval\` field would surface the constraint at call-site:

\`\`\`ts
/** Positive finite seconds; values ≤ 0 or non-finite are clamped to 1s. */
syncInterval?: SyncInterval | number
\`\`\`

File: RCTAppleHealthKit/RCTAppleHealthKit.m (hunk 1 above)
Comment (giladna) [reply RCTAppleHealthKit/RCTAppleHealthKit.m]: Fixed in 7e9dc24. Scoped the removal to each supported event name using removeObserver:name:object: so unrelated observers on self are not affected. 
  ▎ stopObserving retains the broad removeObserver:self since full teardown is correct there.

File: index.d.ts (hunk 2 above)
Comment (giladna) [reply index.d.ts]: Fixed. All three invalid-input paths (number, string, wrong type) now consistently fall back to 86400s. JSDoc in index.d.ts, configureBackgroundSync.md, and mobileapp-core updated to match.

File: RCTAppleHealthKit/RCTAppleHealthKit.m [hunk 3]
```diff
@@ -272,9 +290,20 @@ + (NSTimeInterval)syncIntervalFromString:(NSString *)interval {
...
-    NSString *interval = [input objectForKey:@"syncInterval"];
+    // Accepts either a named string alias ('every1hour', …) or a raw NSNumber of seconds.
+    id interval = [input objectForKey:@"syncInterval"];
     if (interval) {
-        NSTimeInterval seconds = [RCTAppleHealthKit syncIntervalFromString:interval];
+        NSTimeInterval seconds;
+        if ([interval isKindOfClass:[NSNumber class]]) {
+            double raw = [interval doubleValue];
+            // Reject NaN/Infinity/zero/negative — fall back to 24h default.
+            seconds = (isfinite(raw) && raw > 0) ? raw : 86400.0;
```
Comment (yairpi165) [inline RCTAppleHealthKit/RCTAppleHealthKit.m]: ❌ **Missing clamp to 1s minimum**

The PR description states: _"Numeric values clamped to MAX(1.0, seconds)"_, but the actual implementation only checks `raw > 0`, which allows sub-second values like `0.001` to pass through as valid.

Fix:
```objc
seconds = (isfinite(raw) && raw > 0) ? MAX(1.0, raw) : 86400.0;
```

File: RCTAppleHealthKit/RCTAppleHealthKit.m [hunk 4]
```diff
@@ -257,12 +259,28 @@ + (NSTimeInterval)syncIntervalFromString:(NSString *)interval {
...
 RCT_EXPORT_METHOD(configureBackgroundSync:(NSDictionary *)input)
 {
     [self _initializeHealthStore];
+
+    // Lazily wire background observers the first time JS calls configureBackgroundSync.
+    // AppDelegate hooks (sourceURL:, RCTJavaScriptDidLoad) don't fire under Expo New Arch —
+    // this is the only guaranteed JS-reachable entry point on this setup.
+    // Uses a BOOL ivar (not dispatch_once) so a nil-bridge first call can retry on the next call.
+    if (!_observersInitialized) {
+        if (self.bridge) {
```
Comment (yairpi165) [inline RCTAppleHealthKit/RCTAppleHealthKit.m]: ⚠️ **Potential race condition on `_observersInitialized`**

`RCT_EXPORT_METHOD` does not guarantee execution on the main thread. If two JS calls to `configureBackgroundSync` are dispatched concurrently before `_observersInitialized` is set to `YES`, both could pass the `if (!_observersInitialized)` check and call `initializeBackgroundObservers:` twice.

Wrap the initialization block with a main-thread dispatch or use `dispatch_once` with a reset mechanism:
```objc
dispatch_async(dispatch_get_main_queue(), ^{
    if (!self->_observersInitialized && self.bridge) {
        [self initializeBackgroundObservers:self.bridge];
        self->_observersInitialized = YES;
    }
});
```

File: RCTAppleHealthKit/RCTAppleHealthKit+Methods_Fitness.m [hunk 5]
```diff
@@ -499,9 +499,15 @@ - (void)fitness_registerObserver:(NSString *)type
     HKSampleType *sampleType;
     if ([type isEqualToString:@"SleepAnalysis"]) {
         sampleType = [HKObjectType categoryTypeForIdentifier:HKCategoryTypeIdentifierSleepAnalysis];
+    } else if ([type isEqualToString:@"MindfulSession"]) {
```
Comment (yairpi165) [inline RCTAppleHealthKit/RCTAppleHealthKit+Methods_Fitness.m]: ⚠️ **`MindfulSession` added as a special case without explanation**

`MindfulSession` is handled with an explicit `else if` branch — implying `quantityTypeFromName` cannot resolve it. This is correct (it's a category type, not a quantity type), but there is no comment explaining why it needs special handling.

Without context, future developers may add other category types (e.g. `HandwashingEvent`, `ToothbrushingEvent`) directly via `quantityTypeFromName`, which will silently return `nil` and skip the observer (now caught by the nil-guard below, but with no indication of the root cause).

Add a comment:
```objc
// MindfulSession is a category type, not a quantity type — must be resolved explicitly.
} else if ([type isEqualToString:@"MindfulSession"]) {
```

File: index.d.ts [hunk 6]
```diff
@@ -936,17 +936,22 @@ declare module 'react-native-health' {
...
 
   export interface BackgroundSyncOptions {
```
Comment (yairpi165) [inline index.d.ts:953]: ❌ **`SyncInterval | number` allows invalid values at the type level**

Adding `| number` to the union type permits `0`, `-1`, `NaN`, and `Infinity` without any TypeScript compile-time error. The runtime validation in native code catches these, but callers inspecting only the type definition have no indication that constraints exist.

Replace the bare `number` with a branded type or at minimum add a JSDoc `@minimum` annotation:
```ts
/**
 * Minimum time between background delta fetches. Default: `'every24hours'`.
 * Pass a named alias or a raw positive finite number of **seconds** (e.g. `60` for 1 minute).
 * Zero, negative, NaN, and Infinity are rejected at runtime and fall back to 86400s.
 */
export type SyncInterval =
  | 'every1hour'
  | 'every6hours'
  | 'every12hours'
  | 'every24hours'
  | 'every48hours'
  | 'everyweek'
  | number // must be a positive finite value; invalid values default to 86400s at runtime
```

**Review by yairpi165 — CHANGES_REQUESTED** [review]
**2 errors, 2 warnings.**

Key issues:
- Native clamp to `MAX(1.0, seconds)` mentioned in the PR description is missing from the implementation
- `SyncInterval | number` in `index.d.ts` provides no compile-time protection against invalid values
- Race condition risk on `_observersInitialized` without thread-safety guarantee
- `MindfulSession` special case lacks explanation, creating a maintenance trap for future category types

File: RCTAppleHealthKit/RCTAppleHealthKit.m (hunk 3 above)
Comment (giladna) [reply RCTAppleHealthKit/RCTAppleHealthKit.m]: Fixed in dc066c6. Floats are now rounded to nearest integer via MAX(1.0, round(raw)) — e.g. 1.5 → 2, 0.3 → 1. Zero, negative, NaN, and Infinity fall back to 86400s. Updated docs, JSDoc, and mobileapp-core to match.

File: RCTAppleHealthKit/RCTAppleHealthKit.m (hunk 4 above)
Comment (giladna) [reply RCTAppleHealthKit/RCTAppleHealthKit.m]: Fixed in c649110. Wrapped the lazy-init block with @synchronized(self) — serializes concurrent calls so both threads can't pass the if (!_observersInitialized) check simultaneously.
Preferred over dispatch_async to main queue to keep initialization synchronous relative to the method body.

File: index.d.ts (hunk 6 above)
Comment (giladna) [reply index.d.ts:953]: Fixed in 78f8738. Added JSDoc on SyncInterval type with full constraint description + inline comment on number. Branded type rejected — it's a breaking change requiring explicit casting at every call-site, with no benefit over clear documentation since TS can't enforce positive-finite-integer natively.

File: RCTAppleHealthKit/RCTAppleHealthKit+Methods_Fitness.m (hunk 5 above)
Comment (giladna) [reply RCTAppleHealthKit/RCTAppleHealthKit+Methods_Fitness.m]: Fixed in 8ce0143. Added block comment explaining that HKCategoryType observers must be resolved explicitly — quantityTypeFromName can't handle them. Inline comment on MindfulSession line confirms why it needs the explicit branch.

File: index.d.ts [hunk 7]
```diff
@@ -936,17 +936,27 @@ declare module 'react-native-health' {
     | 'last6months'
     | 'lastyear'
 
-  /** Minimum time between background delta fetches. Default: 'every24hours'. */
+  /**
```
Comment (yairpi165) [inline index.d.ts]: ❌ **Duplicate JSDoc block on `SyncInterval`**

There are two consecutive `/** ... */` doc blocks immediately above the `export type SyncInterval` declaration. The first (the old one-liner) was not removed when the expanded block was added. TypeScript tooling will only surface the last block; the first is dead and misleading.

Remove the stale block:
```ts
- /** Minimum time between background delta fetches. Default: 'every24hours'. */
/**
 * Minimum time between background delta fetches. Default: `'every24hours'`.
 * Pass a named alias or a raw positive number of **seconds** ...
 */
```

File: RCTAppleHealthKit/RCTAppleHealthKit.m [hunk 8]
```diff
@@ -257,12 +259,30 @@ + (NSTimeInterval)syncIntervalFromString:(NSString *)interval {
...
+    // Lazily wire background observers the first time JS calls configureBackgroundSync.
+    // AppDelegate hooks (sourceURL:, RCTJavaScriptDidLoad) don't fire under Expo New Arch —
+    // this is the only guaranteed JS-reachable entry point on this setup.
+    // Uses a BOOL ivar (not dispatch_once) so a nil-bridge first call can retry on the next call.
+    @synchronized(self) {
+        if (!_observersInitialized) {
+            if (self.bridge) {
+                NSLog(@"[HealthKit] configureBackgroundSync — lazily initializing background observers");
+                [self initializeBackgroundObservers:self.bridge];
+                _observersInitialized = YES;
```
Comment (yairpi165) [inline RCTAppleHealthKit/RCTAppleHealthKit.m]: ⚠️ **`@synchronized(self)` locks the entire module instance**

`@synchronized(self)` uses the module instance itself as the lock object, which means any concurrent `RCT_EXPORT_METHOD` call on this module will block while the lazy-init runs. For a one-shot guard this is acceptable, but it creates unnecessary contention on subsequent calls (after `_observersInitialized = YES` the lock is still acquired and released on every `configureBackgroundSync` invocation).

Prefer a dedicated lightweight lock that is scoped only to this guard:

```objc
// In the ivar block:
os_unfair_lock _initLock;

// In configureBackgroundSync:
os_unfair_lock_lock(&_initLock);
if (!_observersInitialized && self.bridge) {
    [self initializeBackgroundObservers:self.bridge];
    _observersInitialized = YES;
}
os_unfair_lock_unlock(&_initLock);
```

`os_unfair_lock` is available on iOS 10+ (well within HealthKit baseline) and has no self-deadlock risk here since `initializeBackgroundObservers:` does not re-enter `configureBackgroundSync`.

**Review by yairpi165 — CHANGES_REQUESTED** [review]
All 3 previous comments resolved. 2 new findings:

- Stale JSDoc block left above `SyncInterval` in `index.d.ts` — must be removed (dead doc, misleads tooling)
- `@synchronized(self)` locks the full module instance on every `configureBackgroundSync` call; prefer a dedicated `os_unfair_lock` ivar scoped to this guard

File: RCTAppleHealthKit/RCTAppleHealthKit.m (hunk 1 above)
Comment (giladna) [reply RCTAppleHealthKit/RCTAppleHealthKit.m]: Already implemented this way — `startObserving` at line 1136 already does `removeObserver:self name:notificationName object:nil` in a loop over `supportedEvents`, not the broad `removeObserver:self`. No change needed.

File: index.d.ts (hunk 2 above)
Comment (giladna) [reply index.d.ts]: Added — `BackgroundSyncOptions.syncInterval` already has JSDoc: `/** Positive seconds; floats rounded to nearest integer (min 1s). Zero, negative, NaN, or Infinity default to 86400s (24 hours). */`. Also updated the `| number` inline comment on `SyncInterval` to make constraints explicit.

File: RCTAppleHealthKit/RCTAppleHealthKit.m (hunk 3 above)
Comment (giladna) [reply RCTAppleHealthKit/RCTAppleHealthKit.m]: Already present — line 303: `seconds = (isfinite(raw) && raw > 0) ? MAX(1.0, round(raw)) : 86400.0;`. `MAX(1.0, ...)` clamps sub-second values; `round()` normalizes floats. No change needed.

File: RCTAppleHealthKit/RCTAppleHealthKit.m (hunk 4 above)
Comment (giladna) [reply RCTAppleHealthKit/RCTAppleHealthKit.m]: Fixed — replaced `@synchronized(self)` with `os_unfair_lock` (per yairpi165's comment below). Chose `os_unfair_lock` over `dispatch_async(main_queue)` because: (1) avoids async complexity and the nil-bridge retry still works synchronously, (2) lock is scoped only to the init guard so no contention on unrelated module methods, (3) available iOS 10+ which is well within our baseline.

File: RCTAppleHealthKit/RCTAppleHealthKit+Methods_Fitness.m (hunk 5 above)
Comment (giladna) [reply RCTAppleHealthKit/RCTAppleHealthKit+Methods_Fitness.m]: Already present — lines 500–502 have the block comment explaining the pattern for all HKCategoryType observers, and line 505 has the inline `// HKCategoryType — not resolvable via quantityTypeFromName` note. No change needed.

File: index.d.ts (hunk 7 above)
Comment (giladna) [reply index.d.ts]: Fixed in commit 377cefb — removed stale one-liner, single expanded JSDoc block remains.

File: RCTAppleHealthKit/RCTAppleHealthKit.m (hunk 8 above)
Comment (giladna) [reply RCTAppleHealthKit/RCTAppleHealthKit.m]: Fixed — replaced @synchronized(self) with os_unfair_lock _initLock (private ivar, #import <os/lock.h>). Lock is now scoped only to the init guard, so no contention on unrelated module methods after _observersInitialized = YES.

Kept nested if (self.bridge) (vs your single && condition) to preserve the nil-bridge warning log — behavior is identical.

**Review by yairpi165 — APPROVED** [review]
All previous comments resolved. LGTM ✅
